### PR TITLE
Initial Diagnose Command for TLS and Listener [VAULT-1896, VAULT-1899]

### PR DESCRIPTION
**Pull request by HridoyRoy**
This PR contains the framework for running Diagnose to check TLS and Listener functionality. Some of the test methods are stubs to fill in later, and there's an ActiveProbe stub method for use in phase 2 of Diagnose. 

I also added a helper method in core to fetch the unsanitized vault config, as the sanitized config returns a map[string]interface and not a struct.

### CODE REVIEW

File: command/server.go
Comment (HridoyRoy) [inline command/server.go:959]: // Question: Do I need to return a pointer to the listeners to properly close the listeners in the calling function?

File: vault/core_util.go
Comment (HridoyRoy) [inline vault/core_util.go:62]: I added this initially because I was parsing listeners from core. I removed that functionality from Diagnose, but I think this helper method should still go in for later use.

File: vault/diagnose/diagnose_tls.go
Comment (HridoyRoy) [inline vault/diagnose/diagnose_tls.go:47]: Kept the activeProbe method as a stub, just as a reminder that we should add this into phase 2. I'll make a Jira and link here as well.

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [inline vault/diagnose/diagnose_tls_test.go:134]: Everything below this are stub tests. All this functionality should exist, but I'm having trouble with creating appropriate certs to check these. For example, I'm not able to get an expired TLS cert for TestTLSExpiredCert. I was hoping we could leave these as stubs for now (so we have the Diagnose framework in), but I'm happy to fill these in in this PR as well!

File: command/server/test-fixtures/nostore_config.hcl
Comment (HridoyRoy) [inline command/server/test-fixtures/nostore_config.hcl:7]: 443 is a privileged port, which is why I needed something >= 1024

File: command/server/test-fixtures/tls_config_ok.hcl
Comment (HridoyRoy) [inline command/server/test-fixtures/tls_config_ok.hcl:7]: If I change this to 1024, I get "port already in use" when running the test. This is slightly concerning, since it may point to the fact that the listener may not be closed in the previous runs, but I'm hoping it's just because of the parallelism when running tests.

File: command/operator_diagnose_test.go
Comment (HridoyRoy) [inline command/operator_diagnose_test.go:24]: These parallels cause listeners to collide with one another if they share ports (eg 2 tests using port 1024 will cause a race failure in which either test will randomly fail with a "port already in use" error). So I removed them, allowing listeners to be cleaned up between runs via the defer listener.close block.

Comment (sgmiller) [reply command/operator_diagnose_test.go:24]: Could you assign them each a different incrementing port #?

Comment (sgmiller) [reply command/operator_diagnose_test.go:24]: Oh, possibly not since we don't control the port #s

File: command/server.go
Comment (sgmiller) [reply command/server.go:958]: I would imagine if the listener contains a bound net.Listener, it's pointer would be copied by value.  You can test this though with nestat -tl and a breakpoint before and after you close them.

File: vault/core_util.go
Comment (sgmiller) [reply vault/core_util.go:62]: Does this and the next function repeatedly load the config?  Or is rawConfig set only once?

File: command/server.go
Comment (sgmiller) [reply command/server.go:888]: It looks like it's like an exit code.  Perhaps this function shouldn't use c.UI.Error, but return an error, and the Run form would then display it via c.UI.Error, while the diagnose tests wouldn't need to do so.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:880]: Good point, I can remove the nil check.

File: vault/diagnose/diagnose_tls.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls.go:12]: I can add a helper for 1 listener, but I think we shouldn't separate out the TLS check for listeners from the ConfigUtil.Listener check. The reason is that most of the TLS verification is done for us via helper methods in the x509 library which rely on creating a TLS config from all the different TLS and http related fields in the listener. 

I don't necessarily want to rebuild that -- instead, I want to use these helper methods (as these are what are being used in the server run command).

File: command/operator_diagnose_test.go
Comment (HridoyRoy) [reply command/operator_diagnose_test.go:52]: Definitely agree. Do you want that in this PR, or can I do that in a separate PR and create a jira for it? I wanted to package that into the task to have Diagnose use a tree structure of templates.

File: command/operator_diagnose_test.go
Comment (HridoyRoy) [reply command/operator_diagnose_test.go:24]: We do control the port numbers, and I could assign different port numbers (initially I did that to bypass this problem), but I changed it to use the same port and remove the parallel. I don't like the idea that these tests are going to open a bunch of listeners at the same time on different ports, and removing parallel fixes that issue. 

Also it seems like using Parallel() in our tests is an anti-pattern that we want to avoid.

File: vault/core_util.go
Comment (HridoyRoy) [reply vault/core_util.go:62]: Only once.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:958]: Oh, ok -- let me try that out. Thanks!

File: command/operator_diagnose_test.go
Comment (sgmiller) [reply command/operator_diagnose_test.go:24]: Parallel is actually good so long as it doesn't change the semantics of the tests, helps CI run more quickly.

File: command/operator_diagnose_test.go
Comment (sgmiller) [reply command/operator_diagnose_test.go:52]: Doing it later is fine.  Just want to have something type safe and machine oriented in the future.

File: vault/core_util.go
Comment (sgmiller) [reply vault/core_util.go:62]: Ok, np then.

File: vault/diagnose/diagnose_tls.go
Comment (sgmiller) [reply vault/diagnose/diagnose_tls.go:12]: Okay, makes sense

File: command/operator_diagnose_test.go
Comment (HridoyRoy) [reply command/operator_diagnose_test.go:24]: I spoke too soon! Parallel and different port numbers may be the way to go here after all!

File: command/operator_diagnose_test.go
Comment (HridoyRoy) [reply command/operator_diagnose_test.go:52]: Awesome! Jira: https://hashicorp.atlassian.net/browse/VAULT-1972

File: vault/diagnose/diagnose_tls.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls.go:12]: Actually upon closer inspection, I think you're absolutely right. A lot of the x509 helper methods only need filePaths, and construct a TLSConfig object from those paths. So, I don't necessarily need the entire listener.Config. I separated out a helper function that just takes generic file paths to the TLS info (just the key and cert for now), and performs generic checks on that.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:880]: Done!

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:888]: Done!

File: command/server.go
Comment (HridoyRoy) [reply command/server.go]: Looks ok!

File: command/operator_diagnose.go
Comment (HridoyRoy) [inline command/operator_diagnose.go:31]: Remove parenthesis

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [inline vault/diagnose/diagnose_tls_test.go:1]: change filename to tls.go instead of diagnose_tls.go

File: vault/diagnose/diagnose_tls.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls.go:55]: Throw an error on the stub in case it gets called.

File: vault/diagnose/diagnose_tls_test.go
Comment (swayne275) [inline vault/diagnose/diagnose_tls_test.go:13]: it doesn't look like this logs or errors, so we don't need `t.Helper()`?

File: vault/diagnose/diagnose_tls_test.go
Comment (swayne275) [inline vault/diagnose/diagnose_tls_test.go:77]: i think it would be helpful to fail with the error here. maybe a `t.Errorf()` with the error included?

File: vault/diagnose/diagnose_tls_test.go
Comment (swayne275) [inline vault/diagnose/diagnose_tls_test.go:101]: is this error "type" included in some library we're using under the hood, or does it need to be a string?

File: vault/diagnose/diagnose_tls_test.go
Comment (swayne275) [inline vault/diagnose/diagnose_tls_test.go:130]: does this error string exist in some library, or must it be hard coded this way?

File: vault/core_util.go
Comment (swayne275) [reply vault/core_util.go:62]: what do you mean by "should only be used internally"? to me that implies internal to the package, which would mean it might be better off as a private/unexported function?

File: command/operator_diagnose.go
Comment (swayne275) [inline command/operator_diagnose.go:123]: does the `5` in your `make()` do anything here? my understanding is you don't include a length for a map. i was wondering if maybe there was some special case where that applies to the length of the slice value, but i tried it out in go playground and that doesn't seem to be the case.

File: command/operator_diagnose.go
Comment (swayne275) [inline command/operator_diagnose.go:159]: why do we specify a capacity here?

File: command/operator_diagnose.go
Comment (swayne275) [inline command/operator_diagnose.go:178]: how helpful is it to specify a capacity here? it appears that the required capacity can be lower than `len(config.Listeners)` (if `ln.Config.TLSDisable`, for example)

File: command/operator_diagnose_test.go
Comment (swayne275) [reply command/operator_diagnose_test.go:52]: cool

File: command/server.go
Comment (swayne275) [inline command/server.go:874]: is the first return parameter a return code? it might be helpful to indicate the intent there

File: command/server.go
Comment (swayne275) [inline command/server.go:880]: this looks problematic. there are multiple returns below that don't look like they go through a `c.reloadFuncsLock.Lock()`

i think even if the program exits shortly after one of these returns, we should still be correct with our locking

**Review by swayne275 — CHANGES_REQUESTED** [review]
There's a lot of good stuff here!

I would like to see some of the comments addressed before I approve though (and the ones we discussed on the call).

I'm mostly concerned about the comment about locking

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls_test.go:13]: Yep! That was my understanding as well.

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls_test.go:101]: It's already of type string.

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls_test.go:130]: It exists in a library.

File: command/operator_diagnose.go
Comment (HridoyRoy) [reply command/operator_diagnose.go:159]: Copied from server.go.

File: vault/diagnose/diagnose_tls_test.go
Comment (swayne275) [reply vault/diagnose/diagnose_tls_test.go:101]: i'm wondering if you could do some error comparison like

`if err == storage.ErrObjectNotExist {
	return nil, nil
}`

rather than directly comparing to a string. like does the library have something like this defined

`var (
	// ErrBucketNotExist indicates that the bucket does not exist.
	ErrBucketNotExist = errors.New("storage: bucket doesn't exist")
	// ErrObjectNotExist indicates that the object does not exist.
	ErrObjectNotExist = errors.New("storage: object doesn't exist")
)`

Comment (swayne275) [reply vault/diagnose/diagnose_tls_test.go:101]: oops that formatting is awful, sorry!

File: vault/diagnose/diagnose_tls_test.go
Comment (swayne275) [reply vault/diagnose/diagnose_tls_test.go:130]: so is that string signifying trailing data not defined somewhere in the library then?

File: command/operator_diagnose.go
Comment (swayne275) [reply command/operator_diagnose.go:159]: hmm i think i'm a little confused. what's the reason for having that capacity here too? is there a technical reason to have it?

File: command/operator_diagnose.go
Comment (HridoyRoy) [reply command/operator_diagnose.go:178]: Same as above -- we should avoid using dynamically sized arrays where possible. In this case, I'm willing to preallocate the listener length, even if we don't necessarily use all the space. It would be more costly I think to append individually.

File: command/operator_diagnose.go
Comment (HridoyRoy) [reply command/operator_diagnose.go:123]: Good catch! I used that for testing to explicitly initialize to 5, and forgot to remove it.

File: command/operator_diagnose.go
Comment (swayne275) [reply command/operator_diagnose.go:178]: is this a pretty costly operation?

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:880]: Good catch! I think this was a bug with the run command as well, up until now. I added a defer unlock to fix this.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go]: Done!

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls_test.go]: Done!

File: command/server.go
Comment (swayne275) [reply command/server.go:881]: nice! i'll mark resolved then

File: command/operator_diagnose.go
Comment (HridoyRoy) [reply command/operator_diagnose.go:178]: It doubles the array size every time and copies all elements into a new array. https://blog.golang.org/slices-intro for more info!

Comment (swayne275) [reply command/operator_diagnose.go:178]: i meant for this code specifically. like how many listeners are typical?

Comment (HridoyRoy) [reply command/operator_diagnose.go:178]: I think either way it's not too big a deal. In the case of 3 listeners, we'd end up with an array of 4 with dynamic sizing, versus 3 in the exact case. In the case of 3 listeners with 1 tls_disable listener, we'd end up with 3 in the exact case versus 2 in the dynamic resizing case. 

So in this particular case where the size of the number of listeners is small, I'd say it's more of an idiomatic/personal preference matter.

Comment (swayne275) [reply command/operator_diagnose.go:178]: okay cool on the personal preference bit. i was just curious why that choice was made

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls_test.go]: It's not defined as a static string that I can import. It's defined like so: `err = asn1.SyntaxError{Msg: "trailing data"}` in pkcs7.go

Comment (swayne275) [reply vault/diagnose/diagnose_tls_test.go]: ah, bummer! so either way you'd still have to have some static string in there. alright!

File: command/operator_diagnose.go
Comment (HridoyRoy) [reply command/operator_diagnose.go:159]: Yeah. I think infoKeys here isn't actually relevant -- I kept it mainly to keep the parity between this and server.go, so I could make this dynamic. But I'd want parity with server.go and preexisting code over necessarily cutting out a few extraneous allocations, which is why I went with this.

File: vault/core_util.go
Comment (HridoyRoy) [reply vault/core_util.go:61]: Yeah, that does seem a bit confusing. I just wanted to signify somehow that using the sanitized config is safer, but I'll remove that part of the comment.

File: vault/diagnose/diagnose_tls_test.go
Comment (HridoyRoy) [reply vault/diagnose/diagnose_tls_test.go]: Yeah, I know what you mean, but in this case I can't do it. The error is `return fail(errors.New("tls: failed to find any PEM data in certificate input"))` in src/crypto/tls/tls.go.

File: vault/core_util.go
Comment (HridoyRoy) [reply vault/core_util.go:61]: Removed! Resolving this conversation.

**Review by sgmiller — APPROVED** [review]

File: command/operator_diagnose.go
Comment (swayne275) [inline command/operator_diagnose.go:205]: I'm not familiar with `c.UI.Output` - is the trailing space intended, or will that be auto added? Basically, is this making us have two spaces between "Diagnose..." and the error?

File: command/server.go
Comment (swayne275) [inline command/server.go:894]: is there any way the value for this key could be uninit? i.e. do you want to check if it exists using the two parameter return form?

File: command/server.go
Comment (swayne275) [inline command/server.go:893]: hmm, why are we checking for nil on `reloadFunc` but then using `c.reloadFunc`?

File: command/server.go
Comment (swayne275) [inline command/server.go:932]: what's the advantage to using `fmt.Sprintf` here rather than just assigning `lnConfig.MaxRequestDuration.String()`?

File: vault/core_util.go
Comment (swayne275) [inline vault/core_util.go:76]: is it possible/likely for `*server.Config` to be nil?

**Review by swayne275 — APPROVED** [review]
I'm approving because these are all questions (not critiques), but I would like to have them answered.

Looks good!

File: command/operator_diagnose.go
Comment (HridoyRoy) [reply command/operator_diagnose.go:205]: The trailing space is intended -- from my testing, it isn't added automatically.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:893]: We don't want to append nil reload functions to `relSlice` which is why we check for nil on reloadFunc. c.reloadFuncs can't be nil because it's initialized to an empty map.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:932]: Probably fine to remove the extra fmt.Sprintf wrapping this, but I didn't want to make changes that weren't necessary for this PR. This code was moved from the run function into a helper method, but it's not new.

File: vault/core_util.go
Comment (HridoyRoy) [reply vault/core_util.go:76]: This just asserts that conf is of type `*server.Config` -- if conf is nil, the above nil check will catch it.

File: command/server.go
Comment (HridoyRoy) [reply command/server.go:894]: Yeah, I think that's a good idea!

File: command/server.go
Comment (swayne275) [reply command/server.go:934]: ah alright. your call then but i think it would be an improvement to simplify and remove the fmt.Sprintf()

whichever way you go, feel free to hit the resolve button!

File: vault/core_util.go
Comment (swayne275) [reply vault/core_util.go:76]: ah okay, is it safe to do the `.Sanitized()` for any realistic case here?

I guess I'm concerned because if we're using this as an assertion on a `*server.Config` and the underlying object isn't `*server.Config`, we will panic

basically what i'm asking is: is this safe? if so, go ahead and mark resolved!

File: command/server.go
Comment (HridoyRoy) [reply command/server.go]: I went ahead and removed the wrapping Sprintf.

File: command/server.go
Comment (swayne275) [reply command/server.go]: great! feel free to mark resolved when you do, and then go ahead with the approval!

**Pull request merged**
